cmd/secrets/secret_helper.go

```diff
@@ -53,105 +81,112 @@ type secretsRequest struct {
 	Secrets []string `json:"secrets"`
 }
 
-// ReadSecrets implements a secrets reader from a directory/mount
-func readSecrets(r io.Reader, w io.Writer, dir string) error {
-	in, err := ioutil.ReadAll(r)
+type backend interface {
+	// secretID has a different format on each secret backend. In the file
+	// backend, it's a path, whereas in the kubernetes secret one it has the
+	// following format: "namespace/name/key".
+	get(secretID string) s.Secret
+}
+
+func decodeSecrets(r io.Reader, w io.Writer, dir string, usePrefixes bool, selector *backendSelector) error {
+	inputSecrets, err := parseInputSecrets(r)
 	if err != nil {
 		return err
 	}
 
+	var decodedSecrets map[string]s.Secret
+	if usePrefixes {
+		decodedSecrets = readSecretsUsingPrefixes(inputSecrets, selector)
+	} else {
+		decodedSecrets = readSecretsFromFile(inputSecrets, dir)
+	}
+
+	return writeDecodedSecrets(w, decodedSecrets)
+}
+
+func parseInputSecrets(r io.Reader) ([]string, error) {
+	in, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+
 	var request secretsRequest
 	err = json.Unmarshal(in, &request)
 	if err != nil {
-		return errors.New("failed to unmarshal json input")
+		return nil, errors.New("failed to unmarshal json input")
 	}
 
 	version := splitVersion(request.Version)
 	compatVersion := splitVersion(s.PayloadVersion)
 	if version[0] != compatVersion[0] {
-		return fmt.Errorf("incompatible protocol version %q", request.Version)
+		return nil, fmt.Errorf("incompatible protocol version %q", request.Version)
 	}
 
 	if len(request.Secrets) == 0 {
-		return errors.New("no secrets listed in input")
+		return nil, errors.New("no secrets listed in input")
 	}
 
-	response := map[string]s.Secret{}
-	for _, name := range request.Secrets {
-		response[name] = readSecret(filepath.Join(dir, name))
-	}
+	return request.Secrets, nil
+}
 
-	out, err := json.Marshal(response)
+func writeDecodedSecrets(w io.Writer, resolvedSecrets map[string]s.Secret) error {
+	out, err := json.Marshal(resolvedSecrets)
 	if err != nil {
 		return err
 	}
+
 	_, err = w.Write(out)
 	return err
 }
 
-func splitVersion(ver string) []string {
-	return strings.SplitN(ver, ".", 2)
-}
+func readSecretsFromFile(secrets []string, dir string) map[string]s.Secret {
+	res := make(map[string]s.Secret)
 
-func readSecret(path string) s.Secret {
-	value, err := readSecretFile(path)
-	var errMsg string
-	if err != nil {
-		errMsg = err.Error()
+	secretBackend := fileBackend{rootPath: dir}
+	for _, secretID := range secrets {
+		res[secretID] = secretBackend.get(secretID)
```

Comment:
If we can have errors that aren't variable-specific, we are not propagating them up. I think we will need to add that in here if we can get into a state like that.

Reply:
I'm not sure what you mean here.